scrapy/core/downloader/middleware.py

```diff
@@ -29,9 +33,9 @@ def _add_middleware(self, mw):
         if hasattr(mw, 'process_exception'):
             self.methods['process_exception'].appendleft(mw.process_exception)
 
-    def download(self, download_func, request, spider):
+    def download(self, download_func: Callable, request: Request, spider: Spider):
```

Comment:
Do you think you can add output types as well?

Reply:
Actually, I'm not sure how much info should I add when some of it can be inferred from code.

Reply:
I think we should be annotating public interfaces even if the data types can be inferred, so that Scrapy users can check their code against types provided by Scrapy.

For internal code it is a judgment call if an annotation is needed or not. I'd say there are 2 reasons to annotate, and 2 reasons not to annotate: 

* (+) if it makes code more readable / easy to understand. For example, sometimes type can be inferred from the code, but you don't want to spend your time doing this, it is faster to read the definition.
* (+) if it is likely to prevent errors. For example, you may have some assumptions on what the function should be returning, and if you add e.g. output type information, the assumption would be checked by mypy, including cases when refactoring happens.
* (-) type-annotated code is more verbose, so there is an argument not to overuse it
* (-) it requires effort to add type annotations, which can be spent elsewhere; if the code piece is not complex and not public, or maybe not toched much, the effort might not be justified